docker-setup/action.yml

```diff
@@ -0,0 +1,71 @@
+name: 'Docker build setup'
+description: 'Docker build setup'
+inputs:
+  # GCP auth
+  GCP_WORKLOAD_IDENTITY_PROVIDER:
+    required: true
+  GCP_SERVICE_ACCOUNT_EMAIL:
+    required: true
+  GCP_DOCKER_ARTIFACT_REPO:
+    required: true
+    description: "GCP GAR repo to authenticate to"
+  # AWS auth
+  AWS_ACCESS_KEY_ID:
+    required: true
+  AWS_SECRET_ACCESS_KEY:
+    required: true
+  AWS_DOCKER_ARTIFACT_REPO:
+    required: true
+    description: "AWS ECR repo to authenticate to"
+  # Optional git credentials to use during the build process
+  # Useful if you need to pull private repos for depdenencies
+  GIT_CREDENTIALS:
+    description: "Optional credentials to pass to git"
+    required: false
+  
+runs:
+  using: "composite"
+  steps:
+    - id: auth
+      name: "Authenticate to Google Cloud"
+      uses: "google-github-actions/auth@dac4e13deb3640f22e3ffe758fd3f95e6e89f712" # pin@v0
+      with:
+        create_credentials_file: false
+        token_format: "access_token"
+        access_token_lifetime: 5400 # setting this to 1.5h since sometimes docker builds (special performance builds etc.) take that long. Default is 1h.
+        workload_identity_provider: ${{ inputs.GCP_WORKLOAD_IDENTITY_PROVIDER }}
+        service_account: ${{ inputs.GCP_SERVICE_ACCOUNT_EMAIL }}
+
+    - name: Login to Google Artifact Registry
+      uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b # pin@v2
+      with:
+        registry: us-west1-docker.pkg.dev
+        username: oauth2accesstoken
+        password: ${{ steps.auth.outputs.access_token }}
+
+    - name: Login to ECR
+      uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b # pin@v2
+      with:
+        registry: ${{ inputs.AWS_DOCKER_ARTIFACT_REPO }}
+        username: ${{ inputs.AWS_ACCESS_KEY_ID }}
+        password: ${{ inputs.AWS_SECRET_ACCESS_KEY }}
+
+    - name: setup docker context for buildx
+      id: buildx-context
+      shell: bash
+      run: docker context create builders
+
+    - name: setup docker buildx
+      uses: docker/setup-buildx-action@15c905b16b06416d2086efa066dd8e3a35cc7f98 # pin v2.4.0
+      with:
+        endpoint: builders
+        version: v0.10.1
+
+    - uses: imjasonh/setup-crane@5146f708a817ea23476677995bf2133943b9be0b # pin@v0.1
+
+    - name: Setup git credentials
+      if: inputs.GIT_CREDENTIALS != ''
+      shell: bash
+      run: |
```

Comment:
Can this be made into its own action?

Reply:
yea probably. one thing to note though is that it gets pretty messy if we nest composite actions (since each would have to checkout the action again, and we run into the checkout overwrite issue). If we separate it out into its own action, we'd have to rely on the caller of this action to do them in sequence: like

```
- uses: aptos-labs/actions/docker-setup
- uses: aptos-labs/actions/set-git-build-credentials
````

Reply:
I think I'll leave it as is for now, so as to keep these actions more opinionated